aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -686,8 +687,13 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> native_decoder_only_multi_head_attent
 //     L: Target sequence length
 //     E: Embedding dimension
 std::tuple<Tensor, Tensor> _scaled_dot_product_attention(
-        const Tensor& query_, const Tensor& key, const Tensor& value,
+        const Tensor& query_, const Tensor& key, const Tensor& value, int64_t num_heads,
         const c10::optional<Tensor>& attn_mask_, double dropout_p, bool need_attn_weights, bool is_causal) {
+    // [TODO] Actually replace with the SDP_Backend logic, Figure out how to not use num_heads
+    if(true){
```

Comment:
this will obviously be changed, I spent some time trying to get around have to use num_heads but I couldn't figure out any way to extract out the information from the nested tensors. This is gross and is bleeding in MHA logic to scaled_dot_prod_attention. idk, maybe make this optional and be fine with its grossness or someone smarter than me figure it out. 

Of course this is all rides on the assumption that an input stacked q,k,v is of shape: `(batch_size x ragged_seq_len x (potentially 3* n_heads * head_dim))` Maybe this is where my flaw lies

Reply:
You can require a higher dimensional input that's already reshaped. 

Notice how q, k and v are [already reshaped](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5126) before passed into [_scaled_dot_product](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5185). The `num_heads` check then because a check of `q.size(-1)`.

Reply:
Yeah I am curious if someone tries to shape their tensors so that it will pass all the heuristics required for running the flash_attn kernel but `USE_FLASH_ATTENTION=0` or something how hard does this make it to fallback to the composite implicit version or how much shape mangling are we gunna need for the composite implicit version